File: Server/src/main/java/demo/com/example/testserver/user/dto/UserDTO.java
package demo.com.example.testserver.user.dto;

import java.math.BigDecimal;
import java.util.Date;
import demo.com.example.testserver.user.model.User; // Import User model

// DTO for User responses - excludes sensitive info like password
public class UserDTO {
    private Integer id;
    private String email;
    private String fullName;
    private String avatar;
    private User.UserRole role;
    private User.UserStatus status;
    private BigDecimal customerPoints;
    private Date createdDate;
    private Date updatedDate;
    // Add other fields as needed, e.g., addresses if you want to include them

    // Constructor to map from User entity
    public UserDTO(User user) {
        this.id = user.getId();
        this.email = user.getEmail();
        this.fullName = user.getFullName();
        this.avatar = user.getAvatar();
        this.role = user.getRole();
        this.status = user.getStatus();
        this.customerPoints = user.getCustomerPoints();
        this.createdDate = user.getCreatedDate();
        this.updatedDate = user.getUpdatedDate();
        // Map other fields if added
    }

    // Default constructor (optional, but good practice)
    public UserDTO() {}

    // Getters (required for JSON serialization)
    public Integer getId() { return id; }
    public String getEmail() { return email; }
    public String getFullName() { return fullName; }
    public String getAvatar() { return avatar; }
    public User.UserRole getRole() { return role; }
    public User.UserStatus getStatus() { return status; }
    public BigDecimal getCustomerPoints() { return customerPoints; }
    public Date getCreatedDate() { return createdDate; }
    public Date getUpdatedDate() { return updatedDate; }

    // Setters (optional, useful for testing or manual creation)
    public void setId(Integer id) { this.id = id; }
    public void setEmail(String email) { this.email = email; }
    public void setFullName(String fullName) { this.fullName = fullName; }
    public void setAvatar(String avatar) { this.avatar = avatar; }
    public void setRole(User.UserRole role) { this.role = role; }
    public void setStatus(User.UserStatus status) { this.status = status; }
    public void setCustomerPoints(BigDecimal customerPoints) { this.customerPoints = customerPoints; }
    public void setCreatedDate(Date createdDate) { this.createdDate = createdDate; }
    public void setUpdatedDate(Date updatedDate) { this.updatedDate = updatedDate; }
}

File: ecomerge/lib/database/models/UserDTO.dart
import 'package:e_commerce_app/database/models/user_model.dart'; // Import User model

// Enum for UserRole - match with values in user_model.dart
enum UserRole {
  quan_tri,  // admin
  khach_hang // user
}

// Enum for UserStatus - match with values in user_model.dart
enum UserStatus {
  kich_hoat,  // active
  khoa        // locked/banned
}

// DTO for User responses - excludes sensitive info like password
class UserDTO {
  int? id;
  String? email;
  String? fullName;
  String? avatar;
  String? role;  // Using string to match User model
  String? status; // Using string to match User model
  double? customerPoints;
  DateTime? createdDate;
  DateTime? updatedDate;

  // Constructor to map from User entity
  UserDTO.fromUser(User user) {
    id = user.id;
    email = user.email;
    fullName = user.fullName;
    avatar = user.avatar;
    role = user.role;
    status = user.status;
    customerPoints = user.customerPoints;
    createdDate = user.createdDate;
    updatedDate = user.updatedDate;
  }

  // Default constructor
  UserDTO({
    this.id,
    this.email,
    this.fullName,
    this.avatar,
    this.role,
    this.status,
    this.customerPoints,
    this.createdDate,
    this.updatedDate,
  });

  // Factory constructor to create from JSON
  factory UserDTO.fromJson(Map<String, dynamic> json) {
    return UserDTO(
      id: json['id'],
      email: json['email'],
      fullName: json['fullName'],
      avatar: json['avatar'],
      role: json['role'],
      status: json['status'],
      customerPoints: json['customerPoints'] != null 
          ? double.parse(json['customerPoints'].toString()) 
          : null,
      createdDate: json['createdDate'] != null 
          ? DateTime.parse(json['createdDate']) 
          : null,
      updatedDate: json['updatedDate'] != null 
          ? DateTime.parse(json['updatedDate']) 
          : null,
    );
  }

  // Convert to JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'email': email,
      'fullName': fullName,
      'avatar': avatar,
      'role': role,
      'status': status,
      'customerPoints': customerPoints,
      'createdDate': createdDate?.toIso8601String(),
      'updatedDate': updatedDate?.toIso8601String(),
    };
  }
  
  // Helper method to get role as enum
  UserRole getRoleEnum() {
    if (role == 'quan_tri') {
      return UserRole.quan_tri;
    } else {
      return UserRole.khach_hang;
    }
  }
  
  // Helper method to get status as enum
  UserStatus getStatusEnum() {
    if (status == 'kich_hoat') {
      return UserStatus.kich_hoat;
    } else {
      return UserStatus.khoa;
    }
  }
}
